packages/devtools/lib/src/debugger/console_area.dart

```diff
@@ -54,19 +59,83 @@ class ConsoleArea implements CoreElementView {
     });
   }
 
+  Position _getDocumentEnd() {
+    final doc = _editor.getDoc();
+    final lastLine = doc.lastLine();
+    return Position(lastLine, doc.getLine(lastLine).length);
+  }
+
   void _append(String text) {
-    // append text
-    _editor
-        .getDoc()
-        .replaceRange(text, Position(_editor.getDoc().lastLine() + 1, 0));
-
-    // scroll to end
-    final int lastLineIndex = _editor.getDoc().lastLine();
-    final String lastLine = _editor.getDoc().getLine(lastLineIndex);
-    _editor.scrollIntoView(lastLineIndex, lastLine.length);
+    final scrollInfo = _editor.getScrollInfo();
+    // If the last line of the console is visible, after appending content we
+    // should scroll so that the last line of the editor is still visible.
+    // TODO(jacobr): add an optional setting to always scroll to the end of the
+    // console when new content is added.
+    final scrollToEnd =
+        scrollInfo.top + scrollInfo.clientHeight >= scrollInfo.height;
+
+    final chunks = decodeAnsiColorEscapeCodes(text, _ansiUp);
+
+    final doc = _editor.getDoc();
+    for (StyledText chunk in chunks) {
+      final startPosition = _getDocumentEnd();
+      doc.replaceRange(chunk.text, startPosition);
+
+      final style = chunk.style;
+      if (style != null) {
+        doc.markText(startPosition, _getDocumentEnd(), css: style);
+      }
+    }
+
+    if (scrollToEnd) {
+      final documentEnd = _getDocumentEnd();
+      _editor.scrollIntoView(documentEnd.line, documentEnd.ch);
+    }
   }
 
-  String getContents() {
-    return _editor.getDoc().getValue();
+  /// This method returns a visualization of the contents of the console
+  /// suitable for integration tests.
+  ///
+  /// TextMarkers in the content are visualized with pseudo HTML+CSS to make it
+  /// easy to ensure that the styling of the content matches what was expected.
+  String getStyledContents() {
```

Comment:
How about `String get styledContents {`?

Reply:
this should stay a method as it is quite slow. changed to String styledContents()